### PR TITLE
[k8s-services chart] Add support for custom resources

### PR DESCRIPTION
**Pull request by rhoboat**
Addresses #81, created from #82.

Todos from comment by @yorinasub17 

> Thanks for your contribution! Apologies for the delay in responding here! I lost track of this and just remembered to look.
> 
> This look like a great escape hatch to have here, but we will definitely want to document that this should only be used for resources that are not supported by the chart so that it isn't abused.
> 
> We can definitely move forward to add this in, once the following two things are added:
> 
> - A template test case
> - Documentation in the README

- [x] template test case
- [x] documentation in the README

### CODE REVIEW

File: charts/k8s-service/values.yaml
Comment (brikis98) [inline charts/k8s-service/values.yaml:504]: Why a separate `enabled` flag? Is it not enough to see if `customObjects` is empty or not?

**Review by brikis98 — COMMENTED** [review]
Should we add some examples/tests for this new behavior?

File: charts/k8s-service/values.yaml
Comment (yorinasub17) [reply charts/k8s-service/values.yaml:504]: This is a general pattern in helm and is best practice due to the way merges happen and how helm sets things from the CLI. It is typically very hard to empty out an object if something already sets it in your pipeline, than it is to set an attribute.

Comment (brikis98) [reply charts/k8s-service/values.yaml:504]: Ohhh, got it, thanks for the explanation.

**Comment by rhoboat** [comment]
I pushed a WIP commit and have some questions!

The intent of this PR is to allow for any Kubernetes resources to be created, is that correct? This is different from a CRD, a CustomResourceDefinition, which is a `kind` of resource. Is that correct?

I renamed `customObjects` to `customResources` because I had a different understanding of what this would support. Is `customObjects` the better name? Seems so?

When testing this, what should the helper function that renders the part of the template return? I started thinking we needed a new helper function (` renderK8SServiceCustomResourcesWithSetValues`), but maybe we don't.

I could use some help understanding intent, and that render question.

**Comment by rhoboat** [comment]
@yorinasub17 or @brikis98 Where would it make the most sense to add some documentation about this new support? Does it make sense [under this section](https://github.com/gruntwork-io/helm-kubernetes-services/blob/af1b30e95c5bdc7f6fceae6955bdb56275271e05/charts/k8s-service/README.md#what-resources-does-this-helm-chart-deploy)? (It doesn't make sense in the top level README.)

File: test/k8s_service_custom_resources_example_test.go
Comment (yorinasub17) [inline test/k8s_service_custom_resources_example_test.go:65]: Note for others: To be explicit, I'll be working on this sometime this week as part of support cycles, as a future PR.

**Review by yorinasub17 — DISMISSED** [review]
Updates LGTM!

> Where would it make the most sense to add some documentation about this new support? Does it make sense under this section? (It doesn't make sense in the top level README.)

As far as docs go, I think having a new section titled `How do I deploy additional resources not managed by the chart?` and explaining the escape hatch makes the most sense to me.

**Comment by yorinasub17** [comment]
Looks like the build failed because this is missing the context. Should pass again once you port over one of the gruntwork contexts changes.

**Comment by rhoboat** [comment]
@yorinasub17 Can you take a look at the documentation? How exactly does a user of this chart make changes to it, do they edit it directly, or do they provide a separate values.yaml with `customResources` defined? Do we also want to mention how to deploy using kubectl apply?

File: README.adoc
Comment (yorinasub17) [inline README.adoc:77]: NIT: some of these are already linked down below in the manage section, which is sort of where they belong in the current format. We can discuss if we should do a format change (I do slightly prefer this format), but we should hold off on making those changes until we have decided this so that our repos can be consistent.

File: charts/k8s-service/README.md
Comment (yorinasub17) [inline charts/k8s-service/README.md:52]: This is a GREAT idea. We should do this in all our repos.

File: charts/k8s-service/README.md
Comment (yorinasub17) [inline charts/k8s-service/README.md:56]: NIT:

```suggestion
You can create custom Kubernetes resources that are not directly managed by the chart within the `customResources` key in the chart. You provide the resource manifests directly as a value under `customResources.resources` and set `customResources.enabled` to `true`. For examples of custom resources, take a look at the examples in [test/fixtures/custom_resources_values.yaml](../../test/fixtures/custom_resources_values.yaml) and [test/fixtures/multiple_custom_resources_values.yaml](../../test/fixtures/multiple_custom_resources_values.yaml).
```

**Review by yorinasub17 — DISMISSED** [review]
LGTM! Only had a few nits.

File: README.adoc
Comment (rhoboat) [reply README.adoc:77]: Okay, is it better to revert adding all the links here, then? So there won't be any direct links to most of the stuff in core-concepts.md?

File: charts/k8s-service/README.md
Comment (rhoboat) [reply charts/k8s-service/README.md:52]: We can do it incrementally. AFAIK I started doing it in aws-service-catalog/tls-scripts, and here. So it's just two places for now.

File: README.adoc
Comment (yorinasub17) [reply README.adoc:77]: Sorry should have been more clear on the suggestion. If you agree with what I said, then we should put all these links in the `Manage` section down below in the document.

File: charts/k8s-service/README.md
Comment (rhoboat) [reply charts/k8s-service/README.md:56]: Thank you! Wondered about the wording here.

File: README.adoc
Comment (rhoboat) [reply README.adoc:77]: If you're not strongly opinionated on this, could we start doing this, where we link to everything in core-concepts? So, keep this change, but remove the duplicated links that are already in `manage`.

Comment (rhoboat) [reply README.adoc:77]: I'll push up a change and see what you think. A bit of reorg.

Comment (yorinasub17) [reply README.adoc:77]: I am not strongly opinionated on whether we should do this format change, but I am strongly opinionated that a format change like this should be a team decision and we should go through our regular process to make those team changes. I care a lot about having a consistent look to our docs and having the change isolated in this repo moves counter to that.

Given that, here is a middle ground proposal:

- Keep this section.
- We remove the manage section.
- Before we merge this, start a discussion somewhere talking about this format change. Either an RFC in `team-product` or a `/raise` in slack. Once the discussion is kicked off, we can merge this without waiting on the results (it should be fairly easy to revert it should we decide against it). The reason to gate the merge on starting the discussion is to make sure we have started the discussion before we forget that we did this.

Does that sound good?

Comment (rhoboat) [reply README.adoc]: b15e3ea

Comment (yorinasub17) [reply README.adoc]: Ah the updates LGTM! Let's merge with that change.

**Review by yorinasub17 — DISMISSED** [review]

**Review by yorinasub17 — APPROVED** [review]

**Comment by rhoboat** [comment]
Though CI was skipped in later commits, verified that tests passed: https://circleci.com/gh/gruntwork-io/helm-kubernetes-services/520

**Pull request merged**
